### ISSUE TITLE
Button animation and reaction

### ISSUE THREAD

**Issue by ChristianMayer** [CometVisu/CometVisu#606]
During "non conventional" interaction the buttons (esp. `switch`) doesn't behave as expected / being used to at major OSes (Windows, Linux, Android).

Actions to compare (esp. with mouse as well as with touch):

- Press button and hold, move out, lift
- Press button and hold, move out, move back in, lift
- same but with scroll in between
- same but ended with `Esc` key
- ...

**Comment by peuter** [comment]
And what is the expected behaviour?
* Press button and hold, move out, lift
   => do not send value
* Press button and hold, move out, move back in, lift
   => send value
* same but with scroll in between
   => do not send value?
* same but ended with Esc key
   => do not send value?

Found an additional behaviour difference for triggers:
* Press button and hold (button down animation => ok) 
* move outside the button (button up animation => missing)
* move back inside the button (button down animation => missing)
* lift  (button up animation => ok)

will fix that too.

**Comment by ChristianMayer** [comment]
The issue is mainly about the animations (and the reaction in the sense of the animation).
I will check to be more precise where I see problems with the current state of the code.

**Comment by ChristianMayer** [comment]
Tested with mouse against current develop (i.e. not #610) with demo config:

- Bind Click to Widget Trigger - clicking on the Widget and not the Button does send but doesn't animate
- Trigger: click, move out, release: when leaving the button it should look unpressed again (it's even worse at the moment: releasing the mouse that button is still looking pressed)
- Any button: click, move out, move in, release: should send, but doesn't
- Trigger: click, move out, move in, release: should look click=pressed - move out=unpressed - move in=pressed, release=unpressed,  but doesn't
- Switch: click, scroll and thus move out, move in, release: should send value but doesn't
- Pushbutton: allways looks pressed, but with label "An" should look unpressed (and during click look pressed and show the label "Aus")
- Multitrigger: doesn't show which state is currently set (pressed). Doesn't animate

On mobile device: I can't comment at the moment as the basics aren't working.
When those are fixed it should be checked whether it feels correct. Especially when the user wants to scroll and hits a button (or bind click on widget) either on purpose (as the intention is to click and not to scroll) or on mistake (as the user wants to scroll and not to click). The correct way to handle such a situation can be seen on native apps (e.g. in the OS config menu)

**Comment by ChristianMayer** [comment]
Might relate to / fix #475

**Comment by peuter** [comment]
TODO-List:
- [ ] 1. Bind Click to Widget Trigger - clicking on the Widget and not the Button does send but doesn't animate
- [x] 2. Trigger: click, move out, release: when leaving the button it should look unpressed again (it's even worse at the moment: releasing the mouse that button is still looking pressed)
- [x] 3. Trigger: click, move out, move in, release: should look click=pressed - move out=unpressed - move in=pressed, release=unpressed, but doesn't
- [ ] 4. Any button: click, move out, move in, release: should send, but doesn't
- [ ] 5. Switch: click, scroll and thus move out, move in, release: should send value but doesn't
- [ ] 6. Pushbutton: allways looks pressed, but with label "An" should look unpressed (and during click look pressed and show the label "Aus")
- [ ] 7. Multitrigger: doesn't show which state is currently set (pressed). Doesn't animate

2.+3. Should be fixed with the current state of #610 , although I must admit that I am not sure if I understood 3. right -> needs to be checked

**Comment by peuter** [comment]
Should be fixed now (tests included). I couldn't reproduce 6. but it might have been fixed by the changes I had done so far.

**Comment by peuter** [comment]
@ChristianMayer 
Can we close this issue now?

**Comment by ChristianMayer** [comment]
It definitely is much better now! :+1: 

To be discussed: Clicking on an unpressed switch doesn't animate, only after releasing the mouse. I'd have expected that:

1. switch unpressed
2. Click and hold -> pressed
3. [move out and in -> unpressed and pressed]
4. Release mouse button -> pressed and send pressed-state on the bus

as well as

1. switch pressed
2. Click and hold -> still pressed
3. [move out and in -> also still pressed]
4. Release mouse button -> unpressed and send unpressed-state on the bus

Argument against: currenty the visible button state is also the bus state.
(To be perfected we'd probably need a third state: deeply pressed. Then the switch button can work as a mechanical one-button-switch)

Currently I'm also testing different mobile devices, first impression is fine but I'm not finished yet

**Comment by peuter** [comment]
As you already mentioned the button state (pressed or not) is completely controlled by the bus state. The user immediately sees that something is wrong, when he clicks a switch and the state does not change, because the backend denied the state change for any reason. Otherwise Visu and backend state would be out of sync. Another point is, that an action (aka command to the backend) is only triggered after a real click (which means after a mouseup on the same widget) in a switch. Animating something on mousedown would suggest that something has happened, which in fact doesn't. 

So the current solution is how it should be done IMHO.

I don't understand what you mean with the deeply pressed state.

**Comment by ChristianMayer** [comment]
I'm thinking of something like a "Druckknopfschalter" like https://asset.re-in.de/isa/160267/c1/-/de/702285_RB_00_FB/Druckknopfschalter.jpg?y=525&align=center 

Here you have unpressed -> deeply pressed -> pressed -> deeply pressed -> unpressed

**Comment by ChristianMayer** [comment]
The slider should jump to the mouse position already on the mouse down and not wait for the mouse up.
Here the slider justifies a different behaviour as a slider is typically used in a ways where the approximate position is clicked and then dragged to the final position (let's call it "click and adjust")

This is true for mouse and for mobile device touch

**Comment by ChristianMayer** [comment]
(Older) iPad, (older) Android tablet and modern Android smartphone are working fine.

The device independent issues above (switch and definitely the slider) are still open, of course

**Comment by peuter** [comment]
Slider problem is fixes with #664. And regarding the switch: as I said, I think the current solution is fine.

Still I don't really understand how a three-state button solved the problem:
Do you mean, we should introduce some kind of intermediate state like: "The switch is truned ON by the user, but this has not been confirmed by the backend". So we have a frontend state and a backend state and the cases when they are in sync:
Front.   |   Back.
---------------------
ON       |     ON
OFF     |     OFF

and the cases when they are not in sync:
ON       |     OFF
OFF     |     ON

Is that what you mean?

**Comment by peuter** [comment]
This is more complicated than it seems on the first look. Basically you need the third state, lets call it "unconfirmed" (because the user asked the backend for a change, which did not happen yet).
If the change is confirmed by the backend with a status update, everything is good. But if not (might be because of network problems or because the switch is on a forced state that does not allow changes), the frontend gets no reaction by the backend. For those cases you need some kind of timer, which removes the "unconfirmed" state (and restores the old state, if necessary. Depends how its implemented) after a defined amout of time, if the expected backend update has not been received.

I remember, that we used those kind of "expected updates" in the smart-home system we are developing at work and removed them again because they did not improve the user experience. I don't really remember the exact problem and these two systems might not be comparable in this regard.

Basically the problem can be broken down to the decision if we want immediate user feedback for his interactions and let the frontend show a state the "real world" hasn't reached yet, with the risk that we have to revert that change
**OR**
We always show the "real world" state and go without the immediate feedback.

I prefer the way its done now, because I think the information that something is going wrong when I click on a switch and nothing happens is very useful. But this indeed is a very personal opinion.

**Comment by ChristianMayer** [comment]
The full communication (in the working case) would go:
User -> CV in the Browser -> Backend -> Bus -> Backend -> CV in the Browser -> User

On any place there could something go wrong.
Anything happening after the CV in the browser is shown to the user by the state update of the button (and in some cases in the new message center).
But there's no feedback that shows the user that the CV in the browser did handle the input. A hanging browser does happen sometimes so that's a real world case and only strictly academic.
Also it's good UI to give the user always immediate feedback.

So implementing it as the "deeply pressed" would be the quick and simple way. 
Doing it as a request-confirmation pattern would need probably a bit more work but might(!) be the more satisfying way to do it. Without any experience with that I don't know where the drawbacks are (I could imaging that the pushbutton would be rather strange as this is a write-only widget...).

**Comment by peuter** [comment]
Well the basic problem has to be solved for any widget that allows user interaction and shows backend states. The slider for example needs another, even more complicated solution. The pushbutton actually shows a state (so its not write-only) but needs another solution, and so on and on and on....

Would be great to find a pattern that works with all widgets that need this handling. But whatever we create here, needs a lot of real-world testing, just to see if it "feels good".

**Comment by ChristianMayer** [comment]
As this is at least mostly fine for the moment and a further improvement requires bigger changes on multiple places: postpone

**Comment by ChristianMayer** [comment]
As this is at least mostly fine for the moment and a further improvement requires bigger changes on multiple places: postpone after 0.12.0

**Comment by ChristianMayer** [comment]
Right now no complaints. Closing it. Any new bad observations should open a new issue
